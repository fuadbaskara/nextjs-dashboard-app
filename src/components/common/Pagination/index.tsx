import React, { useEffect, useState } from 'react';
import clsx from 'clsx';
import { AiOutlineLeft, AiOutlineRight } from 'react-icons/ai';
import Button from '@/components/common/Button';

export type PaginationProps = {
  current: number;
  onChange: (page: number) => void;
};

function Pagination({ current, onChange }: PaginationProps) {
  const [pageNumbers, setPageNumbers] = useState<number[]>([1, 2, 3, 4, 5]);
  const [selectedPage, setSelectedPage] = useState<number>(1);

  const onClickPageNumber = (page: number) => {
    setSelectedPage(page);
  };

  const onClickPrev = () => {
    const newPageNumbers = pageNumbers;
    if (current > 1) {
      if (current === pageNumbers[0]) {
        newPageNumbers.unshift(current - 1);
        newPageNumbers.pop();
        setPageNumbers(newPageNumbers);
        setSelectedPage(current - 1);
      } else setSelectedPage(current - 1);
    }
  };

  const onClickNext = () => {
    const newPageNumbers = pageNumbers;
    if (current >= 1) {
      if (current === pageNumbers[pageNumbers.length - 1]) {
        newPageNumbers.push(current + 1);
        newPageNumbers.shift();
        setPageNumbers(newPageNumbers);
        setSelectedPage(current + 1);
      } else setSelectedPage(current + 1);
    }
  };

  useEffect(() => {
    onChange(selectedPage);
  }, [onChange, selectedPage]);

  return (
    <ul className="flex">
      <li>
        <Button
          className="inline-flex items-center rounded-l-md border border-blue-300 bg-white px-2 py-2 text-sm font-medium text-blue-500 hover:bg-blue-50"
          style={{ fontSize: '1.03rem' }}
          type="button"
          onClick={onClickPrev}
        >
          <AiOutlineLeft />
        </Button>
      </li>
      {pageNumbers.map((page) => (
        <li key={page}>
          <Button
            onClick={() => onClickPageNumber(page)}
            className={clsx(
              page === current
                ? 'z-10 border-blue-500 bg-blue-100 text-blue-600'
                : 'border-blue-300 bg-white text-blue-500 hover:bg-blue-50',
              'inline-flex items-center border px-4 py-1.5 text-sm font-medium',
            )}
          >
            {page}
          </Button>
        </li>
      ))}
      <li>
        <Button
          className="inline-flex items-center rounded-r-md border border-blue-300 bg-white px-2 py-2 text-sm font-medium text-blue-500 hover:bg-blue-50"
          style={{ fontSize: '1.03rem' }}
          type="button"
          onClick={onClickNext}
        >
          <AiOutlineRight />
        </Button>
      </li>
    </ul>
  );
}

export default Pagination;
