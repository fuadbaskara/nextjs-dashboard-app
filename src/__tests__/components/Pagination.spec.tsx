/* eslint-disable testing-library/no-node-access */
import React from 'react';
import { screen } from '@testing-library/react';
import { render } from '../test-utils';
import Pagination from '@/components/common/Pagination';

const renderSearch = () =>
  render(
    <Pagination current={1} onChange={(currPage) => jest.fn(() => currPage)} />,
  );

describe('Layout Wrapper Component', () => {
  test('renders correctly with simple props:', () => {
    const { asFragment } = renderSearch();
    const paginationBtn1 = screen.getByTestId('page-1-btn');
    const paginationBtn5 = screen.getByTestId('page-5-btn');
    const paginationPrevBtn = screen.getByTestId('prev-page-btn');
    const paginationNextBtn = screen.getByTestId('next-page-btn');
    expect(paginationBtn1).toBeInTheDocument();
    expect(paginationBtn5).toBeInTheDocument();
    expect(paginationPrevBtn).toBeInTheDocument();
    expect(paginationNextBtn).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
