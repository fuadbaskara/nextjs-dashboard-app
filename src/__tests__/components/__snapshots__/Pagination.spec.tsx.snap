// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout Wrapper Component renders correctly with simple props: 1`] = `
<DocumentFragment>
  <ul
    class="flex"
  >
    <li>
      <button
        class="inline-flex items-center rounded-l-md border border-gray-300 bg-white px-2 py-2 text-sm font-medium text-blue-500 hover:bg-blue-50"
        data-testid="prev-page-btn"
        name="prev-btn"
        style="font-size: 1.03rem;"
        type="button"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 1024 1024"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 0 0 0 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"
          />
        </svg>
      </button>
    </li>
    <li>
      <button
        class="z-10 border-blue-500 bg-blue-100 text-blue-600 inline-flex items-center border px-4 py-1.5 text-sm font-medium"
        data-testid="page-1-btn"
        name="page-1-btn"
        type="button"
      >
        1
      </button>
    </li>
    <li>
      <button
        class="border-gray-300 bg-white text-blue-500 hover:bg-blue-50 inline-flex items-center border px-4 py-1.5 text-sm font-medium"
        data-testid="page-2-btn"
        name="page-2-btn"
        type="button"
      >
        2
      </button>
    </li>
    <li>
      <button
        class="border-gray-300 bg-white text-blue-500 hover:bg-blue-50 inline-flex items-center border px-4 py-1.5 text-sm font-medium"
        data-testid="page-3-btn"
        name="page-3-btn"
        type="button"
      >
        3
      </button>
    </li>
    <li>
      <button
        class="border-gray-300 bg-white text-blue-500 hover:bg-blue-50 inline-flex items-center border px-4 py-1.5 text-sm font-medium"
        data-testid="page-4-btn"
        name="page-4-btn"
        type="button"
      >
        4
      </button>
    </li>
    <li>
      <button
        class="border-gray-300 bg-white text-blue-500 hover:bg-blue-50 inline-flex items-center border px-4 py-1.5 text-sm font-medium"
        data-testid="page-5-btn"
        name="page-5-btn"
        type="button"
      >
        5
      </button>
    </li>
    <li>
      <button
        class="inline-flex items-center rounded-r-md border border-gray-300 bg-white px-2 py-2 text-sm font-medium text-blue-500 hover:bg-blue-50"
        data-testid="next-page-btn"
        name="next-btn"
        style="font-size: 1.03rem;"
        type="button"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 1024 1024"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M765.7 486.8L314.9 134.7A7.97 7.97 0 0 0 302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 0 0 0-50.4z"
          />
        </svg>
      </button>
    </li>
  </ul>
</DocumentFragment>
`;
